tests/devices/test_default_mixed_autograd.py

```diff
@@ -0,0 +1,585 @@
+
+    def test_jacobian_repeated(self, tol):
+        """Test that jacobian of a QNode with an attached default.qubit.autograd device
+        gives the correct result in the case of repeated parameters"""
+        x = 0.43316321
+        y = 0.2162158
+        z = 0.75110998
+        p = np.array([x, y, z], requires_grad=True)
+        dev = qml.device("default.mixed", wires=1)
+
+        @qml.qnode(dev, interface="autograd", diff_method="backprop")
+        def circuit(x):
+            qml.RX(x[1], wires=0)
+            qml.Rot(x[0], x[1], x[2], wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        res = circuit(p)
+
+        expected = np.cos(y) ** 2 - np.sin(x) * np.sin(y) ** 2
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        grad_fn = qml.jacobian(circuit, 0)
+        res = grad_fn(p)
+
+        expected = np.array(
+            [-np.cos(x) * np.sin(y) ** 2, -2 * (np.sin(x) + 1) * np.sin(y) * np.cos(y), 0]
+        )
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    def test_jacobian_agrees_backprop_parameter_shift(self, tol):
+        """Test that jacobian of a QNode with an attached default.qubit.autograd device
+        gives the correct result with respect to the parameter-shift method"""
+        p = np.array([0.43316321, 0.2162158, 0.75110998, 0.94714242], requires_grad=True)
+
+        def circuit(x):
+            for i in range(0, len(p), 2):
+                qml.RX(x[i], wires=0)
+                qml.RY(x[i + 1], wires=1)
+            for i in range(2):
+                qml.CNOT(wires=[i, i + 1])
+            return qml.expval(qml.PauliZ(0)), qml.var(qml.PauliZ(1))
+
+        dev1 = qml.device("default.mixed", wires=3)
+        dev2 = qml.device("default.mixed", wires=3)
+
+        circuit1 = qml.QNode(circuit, dev1, diff_method="backprop", interface="autograd")
+        circuit2 = qml.QNode(circuit, dev2, diff_method="parameter-shift")
+
+        assert circuit1.gradient_fn == "backprop"
+        assert circuit2.gradient_fn is qml.gradients.param_shift
+
+        res = circuit1(p)
+
+        assert np.allclose(res, circuit2(p), atol=tol, rtol=0)
+
+        grad_fn = qml.jacobian(circuit1, 0)
+        res = grad_fn(p)
+        assert np.allclose(res, qml.jacobian(circuit2)(p), atol=tol, rtol=0)
+
+    def test_state_differentiability(self, tol):
+        """Test that the device state can be differentiated"""
+        dev = qml.device("default.mixed", wires=1)
+
+        @qml.qnode(dev, diff_method="backprop", interface="autograd")
+        def circuit(a):
+            qml.RY(a, wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        a = np.array(0.54, requires_grad=True)
+
+        def cost(a):
+            """A function of the device quantum state, as a function
+            of input QNode parameters."""
+            circuit(a)
+            res = np.abs(dev.state) ** 2
+            return res[1][1] - res[0][0]
```

Comment:
This is a great test case, though maybe it's best to return `qml.state()` from the QNode directly. That pipeline should use `dev.state` too under the hood.

Would also be cool to test a QNode that returns `qml.density_matrix(wires=wires)` for some `wires`.

Reply:
Changed the QNode to directly return the state, and also added another test for `qml.density_matrix`.